<a href="https://colab.research.google.com/github/Segun12/HDSC21-Quiz/blob/main/STAGE_B_HAMOYE_QUIZ_BY_OLANIYAN_OLUWASEGUN_EMMANUEL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### To answer some questions, you will need to normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set.

In [1]:
# import library 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

%matplotlib inline

In [2]:
# loading the dataset
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/00374/energydata_complete.csv')

In [4]:
# top 5 rows of the data
df.head()

date  Appliances  lights  ...  Tdewpoint        rv1        rv2
0  2016-01-11 17:00:00          60      30  ...        5.3  13.275433  13.275433
1  2016-01-11 17:10:00          60      30  ...        5.2  18.606195  18.606195
2  2016-01-11 17:20:00          50      30  ...        5.1  28.642668  28.642668
3  2016-01-11 17:30:00          50      40  ...        5.0  45.410389  45.410389
4  2016-01-11 17:40:00          60      40  ...        4.9  10.084097  10.084097

[5 rows x 29 columns]

In [5]:
# drop date and lights features

df_drop = df.drop(['date','lights'], axis = 1)

scaler = MinMaxScaler()
df_scaler = pd.DataFrame(scaler.fit_transform(df_drop), columns=df_drop.columns)

In [6]:
df_scaler.head()

Appliances       T1      RH_1  ...  Tdewpoint       rv1       rv2
0    0.046729  0.32735  0.566187  ...   0.538462  0.265449  0.265449
1    0.046729  0.32735  0.541326  ...   0.533937  0.372083  0.372083
2    0.037383  0.32735  0.530502  ...   0.529412  0.572848  0.572848
3    0.037383  0.32735  0.524080  ...   0.524887  0.908261  0.908261
4    0.046729  0.32735  0.531419  ...   0.520362  0.201611  0.201611

[5 rows x 27 columns]

In [15]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error,r2_score

# Q12

In [42]:
li_reg = LinearRegression()

In [43]:
X1 = df_scaler['T2'].values
y1 = df_scaler['T6'].values

In [44]:
X_train,X_test,y_train,y_test = train_test_split(X1,y1,test_size=0.3, random_state=42)

In [45]:
X_train = X_train.reshape(-1,1)
y_train = y_train.reshape(-1,1)

In [46]:
li_reg.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [47]:
X_test = X_test.reshape(-1,1)

pred = li_reg.predict(X_test)
print('R2_Score in 2 D.P: {:.2f}'.format(r2_score(y_test,pred)))

R2_Score in 2 D.P: 0.64


## Since we were not told to use Q12 to answer all question we will continue with the initial dataset

In [48]:
Xtrain = df_scaler.drop('Appliances', axis=1)
target = df_scaler['Appliances']

In [49]:
X_train,X_test,y_train,y_test = train_test_split(Xtrain,target,test_size=0.3, random_state=42)

li_reg.fit(X_train,y_train)
linear_pred = li_reg.predict(X_test)

In [50]:
print('Q13. R2_Score in 2 D.P: {:.2f} \n Q14. Residual Sum of Square in 2 D.P.: {:.2f} \n Q15. Root Mean Square Error in 3 D.P: {:.3f} \n Q16. Coefficient of Determination  in 2 D.P: {:.2f}'.format(mean_absolute_error(y_test,linear_pred),
                                                                                                                                np.sum(np.square(y_test - linear_pred)),
                                                                                                                                np.sqrt(mean_squared_error(y_test,linear_pred)),
                                                                                                                                r2_score(y_test,linear_pred)))

Q13. R2_Score in 2 D.P: 0.05 
 Q14. Residual Sum of Square in 2 D.P.: 45.35 
 Q15. Root Mean Square Error in 3 D.P: 0.088 
 Q16. Coefficient of Determination  in 2 D.P: 0.15


# Q17

In [39]:
def get_weights_df(model, feat, col_name):
   
  weights = pd.Series(model.coef_, feat.columns).sort_values()
  weights_df = pd.DataFrame(weights).reset_index()
  weights_df.columns = ['Features', col_name]
  weights_df[col_name].round(3)
  return weights_df

In [51]:
# Ridge regression

from sklearn.linear_model import Ridge
ri_reg = Ridge(alpha=0.4)
ri_reg.fit(X_train, y_train)


# Importing the Lasso regression

from sklearn.linear_model import Lasso
la_reg = Lasso(alpha=0.001)
la_reg.fit(X_train, y_train)

Lasso(alpha=0.001, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

In [52]:
linear_model_weights = get_weights_df(li_reg, X_train, 'Linear_Model_Weight')
ridge_weights_df = get_weights_df(ri_reg, X_train, 'Ridge_Weight')
lasso_weights_df = get_weights_df(la_reg, X_train, 'Lasso_weight')

final_weights = pd.merge(linear_model_weights, ridge_weights_df, on='Features')
final_weights = pd.merge(final_weights, lasso_weights_df, on='Features')

In [54]:
# feature weight from Linear model, Ridge model and Lasso model
final_weights

Features  Linear_Model_Weight  Ridge_Weight  Lasso_weight
0          RH_2            -0.456698     -0.411071     -0.000000
1         T_out            -0.321860     -0.262172      0.000000
2            T2            -0.236178     -0.201397      0.000000
3            T9            -0.189941     -0.188916     -0.000000
4          RH_8            -0.157595     -0.156830     -0.000110
5        RH_out            -0.077671     -0.054724     -0.049557
6          RH_7            -0.044614     -0.045977     -0.000000
7          RH_9            -0.039800     -0.041367     -0.000000
8            T5            -0.015657     -0.019853     -0.000000
9            T1            -0.003281     -0.018406      0.000000
10          rv1             0.000770      0.000748     -0.000000
11          rv2             0.000770      0.000748     -0.000000
12  Press_mm_hg             0.006839      0.006584     -0.000000
13           T7             0.010319      0.010098     -0.000000
14   Visibility             0.012307      0.012076      0.000000
15         RH_5             0.016006      0.016152      0.000000
16         RH_4             0.026386      0.024579      0.000000
17           T4             0.028981      0.027384     -0.000000
18    Windspeed             0.029183      0.030268      0.002912
19         RH_6             0.038049      0.035519     -0.000000
20         RH_3             0.096048      0.095135      0.000000
21           T8             0.101995      0.101028      0.000000
22    Tdewpoint             0.117758      0.083128      0.000000
23           T6             0.236425      0.217292      0.000000
24           T3             0.290627      0.288087      0.000000
25         RH_1             0.553547      0.519525      0.017880

In Linear_Model_Weight,

RH_2 has the smallest weight with -0.456698

RH_1 has the highest weight with 0.553547

# Q18

In [55]:
ri_reg.fit(X_train,y_train)

ri_pred = ri_reg.predict(X_test)
print('RMSE: ',np.sqrt(mean_squared_error(y_test,ri_pred)))

RMSE:  0.08753385704628001


# When approximated to 3 D.P there is no difference

# Q19

## From the DataFrame above (final_weights) there 4 non-zero feature weights

# Q20

In [58]:
la_reg.fit(X_train,y_train)

la_pred = la_reg.predict(X_test)
print('RMSE: {:.3f}'.format(np.sqrt(mean_squared_error(y_test,la_pred))))

RMSE: 0.094
